File: Leetcode/540.single-element-in-a-sorted-array.cpp
/*
 * @lc app=leetcode id=540 lang=cpp
 *
 * [540] Single Element in a Sorted Array
 */

// @lc code=start
// //EXPLANATION:-
// Suppose array is [1, 1, 2, 2, 3, 3, 4, 5, 5]
// we can observe that for each pair, 
// first element takes even position and second element takes odd position
// for example, 1 is appeared as a pair,
// so it takes 0 and 1 positions. similarly for all the pairs also.

//this pattern will be missed when single element is appeared in the array.

class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        if(nums.size() == 1){
            return nums[0];
        }
        int right  = nums.size() -1;
        int left = 0;
        while(left < right ){
            
            int mid = (left + right )/2;
            std::cout<<left <<" "<< right<<" "<<mid<<endl;
            if( (mid % 2 == 0 && nums[mid] == nums[mid +1]) || (mid %2 == 1 && nums[mid] == nums[mid - 1]) )
                left = mid + 1;
            else
                right = mid;
        }
           
        return nums[left];
        
    }
};
// @lc code=end

File: Leetcode/540.single-element-in-a-sorted-array.py
#
# @lc app=leetcode id=540 lang=python3
#
# [540] Single Element in a Sorted Array
#

# @lc code=start
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        n = len(nums)
        left = 0
        right = len(nums)-1
        while left < right:
            mid  = left + (right - left)//2
            x1 = nums[mid] == nums[mid+1]
            x2 =  nums[mid] == nums[mid-1]
            if mid > 0:
                if not x1 and not x2 :
                    return nums[mid] 
            else:
                if not x1:
                    return nums[mid]
            print(left, mid, right)
            if nums[mid] != nums[mid+1]:
                if (right-mid)%2 == 0:
                    right = mid
                else:
                    left = mid + 1
            else:
                if (right-mid)%2 == 0:
                    left = mid + 1
                else:
                    right = mid
            print("lastline ", left, mid, right)
        return nums[left]
# @lc code=end

